File: src/main/java/org/example/bankramenserver/domain/recurring/domain/RecurringPayment.java
package org.example.bankramenserver.domain.recurring.domain;

import jakarta.persistence.*;
import lombok.*;
import org.example.bankramenserver.domain.category.domain.Category;
import org.example.bankramenserver.domain.transaction.domain.Transaction;
import org.example.bankramenserver.domain.user.domain.User;

import java.time.LocalDateTime;
import java.time.YearMonth;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor
@Builder
@Table(name = "recurring_payments")
public class RecurringPayment {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(name = "id")
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private Category category;

    @Column(nullable = false)
    private String name;

    @Column(nullable = false)
    private Long amount;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private Cycle cycle;

    @Column(nullable = false)
    private int billingDay;

    @Column(nullable = false)
    private LocalDateTime nextBillingDate;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private RegistrationType registrationType;

    @Column(nullable = false)
    @Builder.Default
    private boolean confirmed = false;

    @Column(nullable = false)
    @Builder.Default
    private boolean active = true;

    @OneToMany(
            mappedBy = "recurringPayment",
            cascade = CascadeType.ALL,
            orphanRemoval = true
    )
    @Builder.Default
    private List<RecurringPaymentTransaction> transactions =
            new ArrayList<>();

    public void confirm() {
        this.confirmed = true;
    }

    public void deactivate() {
        this.active = false;
    }

    public void updateAfterPaymentDetected(
            LocalDateTime nextBillingDate
    ) {
        this.nextBillingDate = nextBillingDate;
    }

    public LocalDateTime calculateNextBillingDate() {

        LocalDateTime base = cycle == Cycle.MONTHLY
                ? nextBillingDate.plusMonths(1)
                : nextBillingDate.plusYears(1);

        YearMonth targetMonth = YearMonth.from(base);

        int actualBillingDay = Math.min(
                billingDay,
                targetMonth.lengthOfMonth()
        );

        return base
                .withDayOfMonth(actualBillingDay)
                .withHour(nextBillingDate.getHour())
                .withMinute(nextBillingDate.getMinute())
                .withSecond(nextBillingDate.getSecond())
                .withNano(nextBillingDate.getNano());
    }

    public void addTransaction(
            Transaction transaction,
            RecurringPaymentTransaction.MatchType matchType,
            LocalDateTime matchedAt
    ) {

        boolean alreadyAdded = transactions.stream()
                .anyMatch(item ->
                        item.getTransaction()
                                .getId()
                                .equals(transaction.getId())
                );

        if (alreadyAdded) {
            return;
        }

        RecurringPaymentTransaction recurringPaymentTransaction =
                RecurringPaymentTransaction.builder()
                        .recurringPayment(this)
                        .transaction(transaction)
                        .matchType(matchType)
                        .matchedAt(matchedAt)
                        .build();

        this.transactions.add(recurringPaymentTransaction);
    }

    public enum Cycle {
        MONTHLY,
        YEARLY
    }

    public enum RegistrationType {
        MANUAL,
        AUTO_DETECTED
    }
}

File: src/main/kotlin/org/example/balogserver/domain/recurring/domain/RecurringPayment.kt
package org.example.balogserver.domain.recurring.domain

import jakarta.persistence.CascadeType
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.OneToMany
import jakarta.persistence.Table
import org.example.balogserver.domain.category.domain.Category
import org.example.balogserver.domain.transaction.domain.Transaction
import org.example.balogserver.domain.user.domain.User
import java.time.LocalDateTime
import java.time.YearMonth
import java.util.UUID

@Entity
@Table(name = "recurring_payments")
class RecurringPayment protected constructor() {
    @field:Id
    @field:GeneratedValue(strategy = GenerationType.UUID)
    @field:Column(name = "id")
    final var id: UUID? = null
        private set
    @field:ManyToOne(fetch = FetchType.LAZY)
    @field:JoinColumn(name = "user_id", nullable = false)
    final lateinit var user: User
        private set
    @field:Enumerated(EnumType.STRING)
    @field:Column(nullable = false)
    final lateinit var category: Category
        private set
    @field:Column(nullable = false)
    final lateinit var name: String
        private set
    @field:Column(nullable = false)
    final var amount: Long? = null
        private set
    @field:Enumerated(EnumType.STRING)
    @field:Column(nullable = false)
    final lateinit var cycle: Cycle
        private set
    @field:Column(nullable = false)
    final var billingDay = 0
        private set
    @field:Column(nullable = false)
    final lateinit var nextBillingDate: LocalDateTime
        private set
    @field:Enumerated(EnumType.STRING)
    @field:Column(nullable = false)
    final lateinit var registrationType: RegistrationType
        private set
    @field:Column(nullable = false)
    final var confirmed = false
        private set
    @field:Column(nullable = false)
    final var active = true
        private set
    @field:OneToMany(mappedBy = "recurringPayment", cascade = [CascadeType.ALL], orphanRemoval = true)
    final var transactions: MutableList<RecurringPaymentTransaction> = ArrayList()
        private set

    constructor(id: UUID?, user: User, category: Category, name: String, amount: Long?, cycle: Cycle, billingDay: Int, nextBillingDate: LocalDateTime, registrationType: RegistrationType, confirmed: Boolean, active: Boolean, transactions: MutableList<RecurringPaymentTransaction>?) : this() {
        this.id = id
        this.user = user
        this.category = category
        this.name = name
        this.amount = amount
        this.cycle = cycle
        this.billingDay = billingDay
        this.nextBillingDate = nextBillingDate
        this.registrationType = registrationType
        this.confirmed = confirmed
        this.active = active
        this.transactions = transactions ?: ArrayList()
    }

    fun confirm() { confirmed = true }
    fun deactivate() { active = false }
    fun updateAfterPaymentDetected(nextBillingDate: LocalDateTime) { this.nextBillingDate = nextBillingDate }
    val isConfirmed get() = confirmed
    val isActive get() = active

    fun calculateNextBillingDate(): LocalDateTime {
        val base = if (cycle == Cycle.MONTHLY) nextBillingDate.plusMonths(1) else nextBillingDate.plusYears(1)
        val actualBillingDay = minOf(billingDay, YearMonth.from(base).lengthOfMonth())
        return base.withDayOfMonth(actualBillingDay)
            .withHour(nextBillingDate.hour)
            .withMinute(nextBillingDate.minute)
            .withSecond(nextBillingDate.second)
            .withNano(nextBillingDate.nano)
    }

    fun addTransaction(transaction: Transaction, matchType: RecurringPaymentTransaction.MatchType, matchedAt: LocalDateTime) {
        if (transactions.any { it.transaction.id!!.equals(transaction.id) }) return
        transactions.add(RecurringPaymentTransaction(this, transaction, matchType, matchedAt))
    }

    companion object { @JvmStatic fun builder() = RecurringPaymentBuilder() }
    class RecurringPaymentBuilder {
        private var id: UUID? = null
        private var user: User? = null
        private var category: Category? = null
        private var name: String? = null
        private var amount: Long? = null
        private var cycle: Cycle? = null
        private var billingDay = 0
        private var nextBillingDate: LocalDateTime? = null
        private var registrationType: RegistrationType? = null
        private var confirmed = false
        private var active = true
        private var transactions: MutableList<RecurringPaymentTransaction>? = null
        fun id(value: UUID?) = apply { id = value }
        fun user(value: User?) = apply { user = value }
        fun category(value: Category?) = apply { category = value }
        fun name(value: String?) = apply { name = value }
        fun amount(value: Long?) = apply { amount = value }
        fun cycle(value: Cycle?) = apply { cycle = value }
        fun billingDay(value: Int) = apply { billingDay = value }
        fun nextBillingDate(value: LocalDateTime?) = apply { nextBillingDate = value }
        fun registrationType(value: RegistrationType?) = apply { registrationType = value }
        fun confirmed(value: Boolean) = apply { confirmed = value }
        fun active(value: Boolean) = apply { active = value }
        fun transactions(value: MutableList<RecurringPaymentTransaction>?) = apply { transactions = value }
        fun build() = RecurringPayment(id, user!!, category!!, name!!, amount, cycle!!, billingDay, nextBillingDate!!, registrationType!!, confirmed, active, transactions)
    }

    enum class Cycle { MONTHLY, YEARLY }
    enum class RegistrationType { MANUAL, AUTO_DETECTED }
}
